helix-core/src/indent.rs

```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
     len / tab_width
 }
 
-/// Find the highest syntax node at position.
-/// This is to identify the column where this node (e.g., an HTML closing tag) ends.
-fn get_highest_syntax_node_at_bytepos(syntax: &Syntax, pos: usize) -> Option<Node> {
-    let tree = syntax.tree();
-
-    // named_descendant
-    let mut node = match tree.root_node().descendant_for_byte_range(pos, pos) {
-        Some(node) => node,
-        None => return None,
-    };
-
-    while let Some(parent) = node.parent() {
-        if parent.start_byte() == node.start_byte() {
-            node = parent
+/// Computes for node and all ancestors whether they are the first node on their line.
+/// The first entry in the return value represents the root node, the last one the node itself
+fn get_first_in_line(mut node: Node, byte_pos: usize, new_line: bool) -> Vec<bool> {
```

Comment:
Isn't it easier to get the first non-whitespace char position in the line (we have a function for that already), then use that with `descendant_for_byte_range`?

Reply:
Yes it would be easier but probably not correct for all grammars. For example we might at some point have a grammar for literate haskell. Because code is marked with a `>` at the beginning of the line, just taking the first non-whitespace char would always (incorrectly) produce the first char.

Reply:
But wouldn't `get_first_in_line` also match this first character?

Reply:
It shouldn't because a reasonable grammar for literate haskell would probably not have `>` as a node in that context and instead ignore it. This enables us to ignore more text than just whitespace if the language requires it.  It might also work better for language injection which we probably want to implement at some point for indent queries.

If you think that this is too specific an issue I can also replace this with simpler approach of just using whitespace.